pam/pam.go

```diff
@@ -73,68 +83,69 @@ func pam_sm_authenticate(pamh *C.pam_handle_t, flags, argc C.int, argv **C.char)
 	p := tea.NewProgram(&appState, opts...)
 	if _, err := p.Run(); err != nil {
 		log.Errorf(context.TODO(), "Cancelled authentication: %v", err)
-		return C.PAM_ABORT
+		return pam.ErrAbort
 	}
 
-	logErrMsg := "unknown"
-	var errCode C.int = C.PAM_SYSTEM_ERR
+	var status = pam.NewTransactionError(pam.ErrSystem, errors.New("unknown exit code"))
 
-	switch exitMsg := appState.exitMsg.(type) {
+	switch exitMsg := appState.exitStatus.(type) {
 	case pamSuccess:
-		brokerIDUsedToAuthenticate = exitMsg.brokerID
-		return C.PAM_SUCCESS
+		if err := mTx.SetData(authenticationBrokerIDKey, exitMsg.brokerID); err != nil {
+			return err
+		}
+		return nil
 	case pamIgnore:
 		// localBrokerID is only set on pamIgnore if the user has chosen local broker.
-		brokerIDUsedToAuthenticate = exitMsg.localBrokerID
-		if exitMsg.String() != "" {
-			log.Debugf(context.TODO(), "Ignoring authd authentication: %s", exitMsg)
-		}
-		logErrMsg = ""
-		errCode = C.PAM_IGNORE
-	case pamAbort:
-		if exitMsg.String() != "" {
-			logErrMsg = fmt.Sprintf("cancelled authentication: %s", exitMsg)
-		}
-		errCode = C.PAM_ABORT
-	case pamAuthError:
-		if exitMsg.String() != "" {
-			logErrMsg = fmt.Sprintf("authentication: %s", exitMsg)
-		}
-		errCode = C.PAM_AUTH_ERR
-	case pamSystemError:
-		if exitMsg.String() != "" {
-			logErrMsg = fmt.Sprintf("system: %s", exitMsg)
+		if err := mTx.SetData(authenticationBrokerIDKey, exitMsg.localBrokerID); err != nil {
+			return err
 		}
-		errCode = C.PAM_SYSTEM_ERR
+		status = exitMsg
+	case pamReturnStatus:
+		status = exitMsg
 	}
 
-	if logErrMsg != "" {
-		fmt.Fprintf(os.Stderr, "Error: %v\n", logErrMsg)
+	if status.Status() != pam.ErrIgnore {
+		if _, err := mTx.StartStringConv(pam.ErrorMsg, status.Error()); err != nil {
+			log.Errorf(context.TODO(), "Failed reporting error to pam: %v", err)
+		}
+	} else if status.Error() != "" {
+		if _, err := mTx.StartStringConv(pam.TextInfo, status.Error()); err != nil {
+			log.Errorf(context.TODO(), "Failed sending info to pam: %v", err)
+		}
 	}
 
-	return errCode
+	return status
 }
 
-// pam_sm_acct_mgmt sets any used brokerID as default for the user.
-//
-//export pam_sm_acct_mgmt
-func pam_sm_acct_mgmt(pamh *C.pam_handle_t, flags, argc C.int, argv **C.char) C.int {
+// AcctMgmt sets any used brokerID as default for the user.
+func (h *pamModule) AcctMgmt(mTx pam.ModuleTransaction, flags pam.Flags,
+	args []string) error {
+	brokerData, err := mTx.GetData(authenticationBrokerIDKey)
```

Comment:
> But that's not possible unless `mTx` is a generic struct too, no?

Yeah, sadly not. The only way to use generics in a method would be to move the type parameter `[T any]` to the struct definition. I don't really know fully the implications of this, so I think it's better to avoid it.